Write phase 3 articles (#8)
CoherentLabs/StarterGuide: src/content/docs/phase-3-layout-assets-and-styling/Graphics & Shapes/image-assets.mdx

---
title: "Image Assets: SVGs vs. Rasters"
description: "Choose between SVG and raster image assets, and load each format in a way that matches its runtime behavior."
sidebar:
    order: 1
---

import Summary from 'coherent-docs-theme/components/Summary.astro';
import Highlight from 'coherent-docs-theme/components/Highlight.astro';
import Link from 'coherent-docs-theme/components/Link.astro';

<Summary>
    Image assets in Gameface should be selected based on the shape of the artwork and the amount of runtime control the UI needs.

    Use <Highlight>SVGs</Highlight> for simple scalable graphics, such as icons, logos, dividers, and HUD frames. Use <Highlight>raster images</Highlight>, such as PNGs, for detailed painted artwork, portraits, screenshots, and textures where vector path evaluation would add unnecessary work.
</Summary>

## The Asset Format Problem 

> In a browser prototype, it is common to pick an image format based mainly on visual quality and file size. In a gameface UI, that decision also affects how much work the UI layer performs during layout, rendering, and invalidation.

Reviewer: I am confused about the `In a gameface UI` what exactly means? Is it the Gameface UI or not? :D

Maybe just `In Gameface` is enough?

Use raster images for artwork that is <Highlight>visually dense</Highlight>: Character portraits, item illustrations, painted backgrounds, screenshots, and textured panels usually belong in PNG, JPEG, or another raster format. 
These assets contain color variation and detail that would require many vector paths to reproduce.

```html title="asset-format-example.html"
<div class="inventory-card">
    <!-- Raster image: high-detail painted artwork -->
    <img class="item-portrait" src="/assets/items/dragon-armor.png" alt="Dragon armor" />

    <!-- SVG image: simple scalable UI symbol -->
    <img class="rarity-icon" src="/assets/icons/legendary-star.svg" alt="Legendary item" />
</div>
```

:::note[The Tradeoffs]
SVGs are easier to scale and recolor, but the renderer still has to process their vector structure. Raster images are less flexible at runtime, but detailed art is usually cheaper to render as a pre-baked pixel image than as a large collection of SVG paths.
:::

### Practical Format Rules 

> For most Gameface UI work, the format decision can be made with a few direct rules:

Reviewer: ```suggestion
Making a decision which format fits the best for your UI in Gameface you can follow these rules:
```

Reviewer: Or something else but different from the original text because it does not make sense written in that way.

* Use **SVG** for icons, simple logos, flat badges, line art, corner decorations, borders, and scalable HUD shapes. And when you need to manipulate the asset at runtime.
* Use **PNG** for transparent high-fidelity UI art, painted panels, portraits, item art, and alpha-heavy textures.
* Use **JPEG** only when the asset does not need transparency and compression artifacts are acceptable, such as some large background images (although the use case is very rare).

:::caution[SVG Usage Gotcha]

> Avoid converting complex painted art into SVG only to make it scalable. If the SVG contains many paths, the visual flexibility may not justify the rendering cost

Reviewer: ```suggestion
Avoid converting complex painted art into SVG only to make it scalable. If the SVG contains many paths, the visual flexibility may not justify the rendering cost.
```

:::

## Working with SVG Assets: External vs. Inline

When using SVGs in Gameface or modern web UIs, you should decide **how** to load them based on your use case. There are two primary strategies for SVG assets: **external SVGs** and **inline SVGs**. Understanding when to use each will help you balance performance, maintainability, and flexibility.

### When to Use External SVGs

If your SVG does **not** need to be manipulated at runtime (for example, recolored or animated by the UI), keep it **external**. This approach minimizes DOM complexity and keeps your structure clean. There are two common external SVG patterns:

#### `<img>` element

> Use when the SVG is meaningful to the content or needs to participate in layout like an image. For example, as a button with an Xbox A glyph icon.

Reviewer: `Use when the SVG is meaningful to the content` - what does this mean? :D

```html title="external-svg-img.html"
<button class="nav-button">
    <img class="nav-button__icon" src="/assets/icons/xbox-a.svg" alt="Xbox A button" />
    <span class="nav-button__label">Select</span>
</button>
```

```css title="external-svg-img.css"
.nav-button {
    display: flex;
    align-items: center;
    gap: 10px;
    height: 56px;
    padding: 0 18px;
}

.nav-button__icon {
    width: 24px;
    height: 24px;
}
```

#### CSS `background-image`

> Use when the SVG is purely decorative and part of the component's styles. For example, a panel frame with a border.

Reviewer: Maybe we can simplify both sections here for the external SVGs. Instead of explaining the differences between img and background-image, we can just list both as valid approaches. The only strict rule we need to highlight is that external SVGs are for graphics that won't be manipulated at runtime, regardless of how they are used.
